Getting grayscale image from camera

alexdtat opened the pull request:
Code for getting image from camera, converting it into grayscale image and displaying it on the screen.

alexdtat committed:
Code for getting image from camera and converting it into grayscale image (now working).

alexdtat commented:
I'm struggling with the architecture. I expect that the code will constantly, in a non-blocking mode, transfer images from the camera to the `ImageAnalysis`, where a grayscale `Plane` will be extracted from it, converted to a `Bitmap` and then displayed. If I understand correctly, I have access to the image only from the analysis function. However: 1) it's not `@Composable`, so `@Composable` functions can't be called inside it, 2) I don't like that the analysing function also displays some content, but I haven't figured out how to avoid it.

P.S. I didn't finish and test  this code because I couldn't compile it, so it's not the final version.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -24,4 +55,100 @@ class MainActivity : ComponentActivity() {
+            height = image.height
+            val buffer = image.planes[0].buffer
+            grayScaleByteArray = buffer.toByteArray()
+            ViewGrayscale(grayScaleByteArray, width, height)
```
I didn't look at the code much as it is WIP and, as I see, the problem lies here.

You can, for example, create a `ViewModel` with `MutableState` fields representing the arguments that are passed here. When created, `ViewGrayscale` will take this `ViewModel` as argument and work with its fields.

TimPushkin replied:
So, here you'll just set these fields to the values you need

alexdtat committed:
Architecture changes: ViewModel for grayscale image has been added.

alexdtat committed:
MutableState instead of MutableLiveData. Delegate fixes. Permissions list fix.

alexdtat committed:
Migration from AppCompatActivity to ComponentActivity. Permissions adaptation.

alexdtat committed:
Conversions fix. Image output changes.

alexdtat committed:
Image rotation fix. Code style fix. Project structure changes.

alexdtat commented:
I hardcoded the 90 deg right rotation because of wrong orientation of images got from camera.

alexdtat committed:
Unused dependencies deletion.

alexdtat commented:
![image](https://user-images.githubusercontent.com/57017816/179833261-e5f59568-8dac-4eb7-941a-c46c769546e6.png)

Current state of app's image proccessing.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/GrayscaleByteArrayToBitmap.kt:
```diff
@@ -0,0 +1,32 @@
+    val byteList = grayscaleByteArray.toList()
+    Log.i(TAG, "Converting ByteArray image of size $width x $height to a Bitmap.")
+    val buffer = IntBuffer.allocate(grayscaleByteArray.size).apply {
+        byteList.forEach { byte ->
```
Here you can do `for (byte in grayscaleByteArray)` and then you won't need to convert it to `List`.

Usually `for (x in y)` is preferable over `y. forEach { x -> ... }` (when possible) anyway as it is a reserved word made for this specifically making the code more readable.

alexdtat replied:
fixed

TimPushkin replied:
You don't need `toList` conversion now, so remove it

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/GrayscaleByteArrayToBitmap.kt:
```diff
@@ -0,0 +1,32 @@
+private const val RED_SHIFT = 16 // Red channel shift
+private const val GREEN_SHIFT = 8 // Green channel shift
+private const val BLUE_SHIFT = 0 // Blue channel shift
+private const val Y_SHIFT = 127 // Y-component should be inverted and shifted
```
It's better to separate this shift from the other (by one empty line) because it's a little different from them

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/GrayscaleViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+
+    @Suppress("FunctionNaming")
+    @Composable
+    fun ViewGrayscale() {
```
`ViewModel` is not a place for composable (or any other UI elements that don't need their state to be holded): you can read more about it [here](https://developer.android.com/topic/architecture#ui-layer).

It is a good practice to create separate files for composables as you do for classes.

TimPushkin replied:
If you think there will be a problem with reading `GrayscaleViewModel`'s fields, I think, there shouldn't be any if you pass them as arguments (you can even pass the whole `ViewModel` but it's not needed here) because they are `MutabelState`s.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/GrayscaleViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+    private var width by mutableStateOf(0)
+    private var height by mutableStateOf(0)
+
+    @Suppress("FunctionNaming")
```
Is this required for some code analysis tool? Android Studio seems to have no problems with the special composables naming.

alexdtat replied:
> Is this required for some code analysis tool? Android Studio seems to have no problems with the special composables naming.

It required for detekt. I planned to add it to the CI, but I'm not sure how useful it will be.

TimPushkin replied:
It'll be useful in case you forget to run it or somebody else (me, for example) commits

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/GrayscaleViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+    @Composable
+    fun ViewGrayscale() {
+        if (width == 0 || height == 0) return
+        val grayscaleBitmap = grayscaleByteArrayToBitmap(
```
I don't think a composable should do such calculations: composables should be as light as possible. Why don't you place it into `setPicture` and just store the resulting `Bitmap` in the `ViewModel`?

alexdtat replied:
> I don't think a composable should do such calculations: composables should be as light as possible. Why don't you place it into `setPicture` and just store the resulting `Bitmap` in the `ViewModel`?

`Bitmap` needs `width` > 0 and `height` > 0 for construction so `mutableStateOf` will have at least 1x1 `Bitmap` image as argument.
UPD: I use nullable varibale. Fixed.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/GrayscaleViewModel.kt:
```diff
@@ -0,0 +1,40 @@
+        Image(
+            bitmap = grayscaleBitmap.asImageBitmap(),
+            modifier = Modifier.fillMaxSize(),
+            contentDescription = "Grayscale photo"
```
Just for your knowledge: in production you are required to localize `contentDescription`s as they are used for [accessibility purposes](https://developer.android.com/jetpack/compose/accessibility#describe-visual).

Leave this as is here, though, because we want to keep the UI code as simple as possible for now.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+        cameraExecutor = Executors.newSingleThreadExecutor()
         setContent {
-            Theme {
+            MaterialTheme {
```
This should be the theme you defined in `Theme.kt` (so, `Theme` here)

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
         super.onCreate(savedInstanceState)
+
+        checkPermissions()
+        cameraExecutor = Executors.newSingleThreadExecutor()
```
Why not place the initialization  where the property is declared without making it `lateinit`? It's not something that needs `Activity` to be initialized.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+                        "Permissions granted.",
+                        Toast.LENGTH_SHORT
+                    ).show()
+                    finish()
```
When I start the app the first time and grant the permissions this makes the app close and I have to reopen it manually again

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+                        "Permissions not granted by the user.",
+                        Toast.LENGTH_SHORT
+                    ).show()
+                    finish()
```
I think a better solution would be to still start the app but display something like "Camera permission required" in place of the image

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+
+    private val cameraPermission =
+        registerForActivityResult(ActivityResultContracts.RequestPermission()) { granted ->
+            when {
```
`if`-`else` would be simpler here

alexdtat replied:
fixed
UPD: Using `when` because of `shouldShowRequestPermissionRationale`.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+        ContextCompat.checkSelfPermission(
+            this,
+            it
+        ) == PackageManager.PERMISSION_GRANTED
```
When `checkSelfPermissions` returns `DENIED` you should call `shouldShowRequestPermissionRationale` for this permission and then probably explain to a user why you need this permission. The workflow is nicely described [here](https://developer.android.com/training/permissions/requesting#request-permission) with a code snippet.

TimPushkin replied:
Maybe you'll be able to simplify your code as you only need one permission

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+
+    private fun checkPermissions() {
+        if (allPermissionsGranted()) {
+            startCamera()
```
It's strange that a function called `checkPermissions` starts the camera don't you think?

alexdtat replied:
Will be `tryStartCamera` suitable function name? Or `try` is too strongly associated with exceptions?

TimPushkin replied:
It's ok

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()
+
+            val imageAnalyzer = ImageAnalysis.Builder()
+                .setOutputImageFormat(OUTPUT_IMAGE_FORMAT_YUV_420_888)
```
[Docs say](https://developer.android.com/reference/androidx/camera/core/ImageAnalysis.Builder#setTargetRotation(int)) we can also set image rotation here. Won't it help with the rotation problem you have?

TimPushkin replied:
There is also an example of getting image rotation in the [code snippet here](https://developer.android.com/training/camerax/analyze#bind-to-lifecycle)

alexdtat replied:
> [Docs say](https://developer.android.com/reference/androidx/camera/core/ImageAnalysis.Builder#setTargetRotation(int)) we can also set image rotation here. Won't it help with the rotation problem you have?

Unfortunately, it seems, this method does nothing in our case (I've tried all possible rotations with the same result). But maybe I missed something.

UPD: A little misunderstanding. Yes, we can use it to improve the hardcoded rotation.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+            val cameraProvider: ProcessCameraProvider = cameraProviderFuture.get()
+
+            val imageAnalyzer = ImageAnalysis.Builder()
+                .setOutputImageFormat(OUTPUT_IMAGE_FORMAT_YUV_420_888)
```
You probably also want to set [an operation mode](https://developer.android.com/training/camerax/analyze#operating-modes) here. Remember we wanted to use non-blocking?

alexdtat replied:
But the documentation says that non-blocking mode is set by default. However, I can do it explicitly.

TimPushkin replied:
Oh, ok than, no need to.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+                .setOutputImageFormat(OUTPUT_IMAGE_FORMAT_YUV_420_888)
+                .build()
+                .also {
+                    it.setAnalyzer(cameraExecutor, PhotoAnalyzer(imageViewModel))
```
Won't `apply` be better here? This `it` will be eliminated this way.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+        }, ContextCompat.getMainExecutor(this))
+    }
+
+    private class PhotoAnalyzer(val imageViewModel: GrayscaleViewModel) : ImageAnalysis.Analyzer {
```
Maybe extract this class to another file?

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+            rewind()
+            val data = ByteArray(remaining())
+            get(data)
+            return data
```
I would write this three lines as `return ByteArray(remaining()).also { get(it) }` but maybe I love scope functions too much...

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+        ) = ByteArray(width * height).apply {
+            for (i in 0 until width) {
+                for (j in 0 until height) {
+                    this[i * height + j] = grayscaleByteArray[i + (height - 1 - j) * width]
```
Replacing `apply` with `also { array -> ...` and `this` with `array` seems better to me

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+            }
+        }
+
+        // Here was hardcoded the 90 deg right rotation because of wrong orientation of images got from camera.
```
It should be "the 90 deg right rotation is hardcoded".

And isn't this line too long?

alexdtat replied:
In detekt max line size is aroud 120 symbols IIRC. Though I can split it.

TimPushkin replied:
100 is the default in Android Studio, let's keep it this way

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+        }
+    }
+
+    companion object {
```
Why have you chosen `companion object` over top-level constants, what are the pros of this? I see the following cons:
1. This creates a separate object associated with the class.
2. This requires more lines of code.

alexdtat replied:
In the past I was advised to use `companion object` for constants. Though I googled a little bit and it seems to be a not welcomed practise.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+    }
+
+    companion object {
+        private const val TAG = "FeatureDetectApp"
```
`TAG` is usually the name of the class. It's just a convention but I think debugging is much easier this way -- you immediately see where to look from the logs.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,153 @@
+
+    companion object {
+        private const val TAG = "FeatureDetectApp"
+        private val REQUIRED_PERMISSIONS = mutableListOf(Manifest.permission.CAMERA).toTypedArray()
```
What do you need `toTypedArray()` for?

alexdtat replied:
Rudimentary code. Fixed.

TimPushkin commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,6 +2,12 @@
+        android:name="android.hardware.camera2"
+        android:required="true"/>
+    <uses-permission android:name="android.permission.CAMERA" />
+    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE"
```
I didn't see this used anywhere

alexdtat replied:
fixed

TimPushkin commented on app/build.gradle:
```diff
@@ -40,9 +40,17 @@ android {
         }
     }
+    buildFeatures {
+        viewBinding true
```
Maybe I missed something, but I didn't see any view binding in the code

alexdtat replied:
fixed

TimPushkin reviewed (changes requested):
Seems great overall!

alexdtat committed:
MainActivity decomposition (PhotoAnalyzer is now in PhotoAnalyzer.kt). Image rotation hardcode improvement. Code style and naming fixes. Rudimentary code deletion.

alexdtat committed:
GrayscaleViewModel decomposition (GrayscaleView.kt is now added). Code style and naming fixes.

TimPushkin commented on app/build.gradle:
```diff
@@ -43,6 +43,11 @@ android {
 }
 
 dependencies {
+    def camerax_version = "1.1.0-beta01"
```
You don't need to use `"..."` here as it's not a formatted string -- `'...'` would be enough

TimPushkin replied:
There are several other places in this file where you can replace double quotes with single ones. Not a big deal, but still.

alexdtat replied:
fixed

TimPushkin commented on app/build.gradle:
```diff
@@ -43,6 +43,11 @@ android {
+    def camerax_version = "1.1.0-beta01"
+    implementation "androidx.camera:camera-core:${camerax_version}"
+    implementation "androidx.camera:camera-camera2:${camerax_version}"
+    implementation "androidx.camera:camera-lifecycle:${camerax_version}"
```
You don't need to use braces in any of these lines

alexdtat replied:
fixed

TimPushkin commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,6 +2,10 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
     package="com.github.featuredetectandroid">
+    <uses-feature
```
If you press `Ctrl+Alt+L` in this file, Android Studio will fix some formatting issues

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/GrayscaleView.kt:
```diff
@@ -0,0 +1,23 @@
+) {
+    if (width == 0 || height == 0) return
+    Image(
+        bitmap = grayscaleBitmap?.asImageBitmap() ?: return,
```
I think, you can just do `grayscaleBitmap?.let { bitmap -> ...` and remove all other checks. If somebody passes image of zero size it's their fault.

TimPushkin replied:
Maybe you don't even need this composable because it is just an `Image` and the `ViewModel` because it only stores the bitmap? Such `ViewModel` could be replaced with `rememberSaveable`, though, it may have problems saving large bitmaps, so I'm not sure.

alexdtat replied:
I inlined `View` and kept `ViewModel`.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/theme/Theme.kt:
```diff
@@ -21,6 +21,7 @@ private val LightColorPalette = lightColors(
 
 @Composable
+@Suppress("FunctionNaming")
 fun Theme(darkTheme: Boolean = isSystemInDarkTheme(), content: @Composable () -> Unit) {
```
Maybe rename this to something more descriptive? Now it looks like something built-in in Jetpack Compose.

alexdtat replied:
> Maybe rename this to something more descriptive? Now it looks like something built-in in Jetpack Compose.

Yes, even I mixed them up last time...

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/GrayscaleArrayToBitmap.kt:
```diff
@@ -0,0 +1,36 @@
+private const val MAX_COLOR = 0xff
+private const val MAX_ALPHA = MAX_COLOR shl ALPHA_SHIFT
+
+fun grayscaleArrayToBitmap(grayscaleByteArray: ByteArray, width: Int, height: Int): Bitmap {
```
Y in YUV is called "luminance", so, as you are expecting it here, it's probably better to replace "grayscale" with "luminance"

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+        return ByteArray(remaining()).also { get(it) }
+    }
+
+    private fun rotateRightOn90Deg(
```
"Right" is not very descriptive when we are talking about rotations. It should be either "clockwise" or "counterclockwise".

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+        var height = image.height
+        val buffer = image.planes[0].buffer
+        val grayScaleByteArray = buffer.toByteArray()
+        val oriented = if (rotationDegrees == ROTATION_ON_90_DEGREES) {
```
`rotationDegrees`'s docs say that it can be any of 0, 90, 180, 270 degrees. As a temporary solution you can apply the 90-degree rotation `rotationDegrees / 90` times and add a `TODO` to fix this later.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+import java.nio.ByteBuffer
+
+private const val TAG = "PhotoAnalyzer"
+private const val ROTATION_ON_90_DEGREES = 90
```
Something like `ROTATION_STEP` would be better

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    /* Here is the 0 or 90 deg (depending on state) right rotation hardcoded
+    because of wrong orientation of some images got from camera. */
```
Sounds not really good in English for me, though, I'm not a native speaker too. It would be better to make this comment shorter and place this where the rotation happens.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,138 @@
+                    if (ContextCompat.checkSelfPermission(
+                            this@MainActivity,
+                            Manifest.permission.CAMERA
+                        ) == PackageManager.PERMISSION_GRANTED
```
It seems like this check is run only when the app starts and isn't rerun after the permission is granted. So, the following happens:
- I start the app the first time
- It asks me for permissions and this check fails
- I grant the permissions
- Camera view doesn't appear because the check is not rerun

And I need to restart the app to see the output :)

alexdtat replied:
Fixed, if I understood you correctrly.

TimPushkin replied:
Nope, this still happens. To reproduce the behaviour do the following:
1. Remove the app's camera permission (or just reinstall it).
2. Start the app.
3. Grant camera permissions.
4. Still see the "Camera permission required" text.

I think, you need to create a `Boolean` field in your `ViewModel` and track permission status there so that Compose sees its updates.

alexdtat replied:
Fixed. Also added switching between windos support.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,138 @@
+                            imageViewModel.height
+                        )
+                    } else {
+                        Text(text = "Camera permission required")
```
This text appears in the corner of the screen. You need to configure center alignment in the `Box`.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,138 @@
+        when {
+            cameraPermissionGranted() -> {
+                startCamera()
+            }
```
Braces are not needed here

alexdtat replied:
> Braces are not needed here

Wouldn't it be better to use braces here for reasons of uniformity?

TimPushkin replied:
I think, it's more readable without additional braces, because it takes less space. Style guide also seems to [permit this](https://developer.android.com/kotlin/style-guide#braces).

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,138 @@
+                startCamera()
+            }
+            rationaleSDKVersionCheck() &&
+                shouldShowRequestPermissionRationale(Manifest.permission.CAMERA) -> {
```
Maybe you can put these two calls into a single function? It'll look better.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,138 @@
+            }
+            else -> {
+                cameraPermission.launch(Manifest.permission.CAMERA)
+            }
```
Braces are not needed here too

alexdtat replied:
fixed

TimPushkin reviewed (changes requested):
More changes required

alexdtat committed:
180 and 270 deg rotations are now supported. Code style and naming fixes.

alexdtat committed:
GrayscaleView function was inlined. First permission issue fix.

alexdtat committed:
Camera resolution setter was added.

alexdtat commented:
https://github.com/TimPushkin/FeatureDetectAndroid/pull/4#discussion_r925884182

The question of using `rememberSaveable` instead of `ViewModel` is still open.

TimPushkin commented:
> The question of using `rememberSaveable` instead of `ViewModel` is still open.

Yeah, seems like you need to keep this `ViewModel`.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,146 @@
+                        this,
+                        "Permissions granted.",
+                        Toast.LENGTH_SHORT
+                    ).show()
```
It can be squeezed into one line

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,146 @@
+                        this,
+                        "Permissions not granted by the user.",
+                        Toast.LENGTH_SHORT
+                    ).show()
```
You can replace "Permissions not granted by the user." with "Permissions not granted." for uniformity and also squeeze it into a single line

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,146 @@
+                        this,
+                        "Permissions not granted by the user.",
+                        Toast.LENGTH_SHORT
+                    ).show()
```
Why do we log that permissions were granted but don't log that they weren't?

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,146 @@
+                        ) == PackageManager.PERMISSION_GRANTED
+                    ) {
+                        Image(
+                            bitmap = imageViewModel.grayscaleBitmap?.asImageBitmap() ?: return@Box,
```
Labeled statements should be avoided when possible because they are hard to understand. Just surround the `Image` with `imageViewModel.grayscaleBitmap?.let { bitmap -> ... }` -- this would look much less cumbersome.

TimPushkin replied:
I thought detekt has warnings about labeled statements

alexdtat replied:
No, as far as I know from my experience, detekt does not prohibit lebeled statements. :)
Fixed.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,146 @@
+        android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M
+
+    private fun tryShowRequestPermissionRationale() = rationaleSDKVersionCheck() &&
+        shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)
```
Unite `rationaleSDKVersionCheck()` and `tryShowRequestPermissionRationale()` into a single function and name it something like `shouldRationalize()`

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,146 @@
+                    this,
+                    cameraSelector,
+                    imageAnalyzer
+                )
```
This can fit into a single line. Personally I only see a point in writing arguments like this in two cases:
1. They don't fit into a single line.
2. Several named arguments are used.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/GrayscaleViewModel.kt:
```diff
@@ -0,0 +1,20 @@
+            grayscaleByteArray,
+            newWidth,
+            newHeight
+        )
```
This can fit into a single line

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/ui/theme/FeatureDetectAppTheme.kt:
```diff
@@ -21,7 +21,11 @@ private val LightColorPalette = lightColors(
 
 @Composable
-fun Theme(darkTheme: Boolean = isSystemInDarkTheme(), content: @Composable () -> Unit) {
+@Suppress("FunctionNaming")
```
I don't have much experience with detekt but we can probably configure it as described [here](https://detekt.dev/docs/introduction/compose/#functionnaming) to remove such supressions

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    // Rotation of incorrectly oriented images is implemented here.
+    @Suppress("ForbiddenComment")
```
Maybe it would be better to put this comment where the rotation happens (inside the function) and get rid of this supression?

alexdtat replied:
> Maybe it would be better to put this comment where the rotation happens (inside the function) and get rid of this supression?

"ForbiddenComment" is comment with "TODO".

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+        var oriented = image.planes[0].buffer.toByteArray()
+
+        // TODO: Optimize rotations for 180 and 270 degrees.
+        repeat((0 until (rotationDegrees / ROTATION_STEP)).count()) {
```
Isn't it the same as `repeat(rotationDegrees / ROTATION_STEP) { ... }`?

alexdtat replied:
Yes, these options are the same. Fixed.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+
+        // TODO: Optimize rotations for 180 and 270 degrees.
+        repeat((0 until (rotationDegrees / ROTATION_STEP)).count()) {
+            oriented = rotateClockwiseOnRotationStep(oriented, width, height).also {
```
Why do you need `also` here?

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,50 @@
+            }
+        }
+
+        imageViewModel.setPicture(oriented, width, height)
```
I'm a bit frustrated, probably you can explain this to me. So, we need to do two transformations: rotation and bitmap conversion. Why one of them happens inside `PhotoAnalyzer` and the other one in `ViewModel`?

alexdtat replied:
> I'm a bit frustrated, probably you can explain this to me. So, we need to do two transformations: rotation and bitmap conversion. Why one of them happens inside `PhotoAnalyzer` and the other one in `ViewModel`?

Rotation when analyzing a photo is needed in order to be able to invoke the functions of the native library from `analyze` with correct images. The `Bitmap` is used only for the GUI, so I decided that it is better to delegate the logic for obtaining it to the `ViewModel`. But, if it's preferable, I can do the `Bitmap` conversion here.

TimPushkin replied:
Ok, this makes sense to me, leave it in the `ViewModel`

TimPushkin reviewed (changes requested):
We're getting close 🙃

alexdtat committed:
First time permission and switching between windows for granting permission fix. Code style and naming fixes. Detekt suppressions elimination from this branch.

alexdtat commented:
I hardcoded the camera resolution to 640x360 and removed suppressions for detekt from this PR. In addition, I added support for rotating images 180 and 270 degrees using 90 degree rotations. In the future, it is preferable to replace this with a more efficient solution.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/MainActivity.kt:
```diff
@@ -1,27 +1,136 @@
+
+    private fun shouldRationalize() =
+        (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) &&
+            shouldShowRequestPermissionRationale(Manifest.permission.CAMERA)
```
If I press `Crtl+Alt+L` Android Studio adds additional indents here

alexdtat replied:
Hm. Ktlint (with `adroid` flag) gives me an error for any other indents.

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/LuminanceArrayToBitmap.kt:
```diff
@@ -0,0 +1,36 @@
+            val intByte = Y_SHIFT - byte.toInt()
+            put(
+                MAX_ALPHA or (intByte shl RED_SHIFT)
+                    or (intByte shl GREEN_SHIFT) or (intByte shl BLUE_SHIFT)
```
[According to Android code style](https://developer.android.com/kotlin/style-guide#where_to_break), line break should come after `or`. Also, if I press `Crtl+Alt+L` Android Studio adds additional indents here.

alexdtat replied:
fixed

TimPushkin commented on app/src/main/java/com/github/featuredetectandroid/utils/PhotoAnalyzer.kt:
```diff
@@ -0,0 +1,48 @@
+        var oriented = image.planes[0].buffer.toByteArray()
+
+        /** Rotation of incorrectly oriented images is implemented here.
+         * TODO: Optimize rotations for 180 and 270 degrees. */
```
Why is this a [KDoc](https://kotlinlang.org/docs/kotlin-doc.html#kdoc-syntax) string? I would just place two comments with `//`.

alexdtat replied:
fixed

TimPushkin reviewed (changes requested):
A few code style changes left and we are done here

alexdtat committed:
Small code style fixes.
